rs/registry/admin/src/main.rs

```diff
@@ -5530,6 +5551,129 @@ async fn propose_to_add_or_remove_node_provider(
+    let install_mode = match cmd.mode {
+        CanisterInstallMode::Install => Some(GovernanceInstallMode::Install as i32),
+        CanisterInstallMode::Reinstall => Some(GovernanceInstallMode::Reinstall as i32),
+        CanisterInstallMode::Upgrade => Some(GovernanceInstallMode::Upgrade as i32),
+    };
+
+    let install_code = InstallCode {
+        skip_stopping_before_installing,
+        install_mode,
+        canister_id,
+        wasm_module,
+        arg,
+    };
+
+    print_proposal(&install_code, &cmd);
+
+    if cmd.is_dry_run() {
+        return;
+    }
+
+    let proposal_id = canister_client
+        .submit_install_code_proposal(install_code, cmd.url(), cmd.title(), cmd.summary())
+        .await
+        .unwrap_or_else(|e| {
+            eprintln!("propose_to_install_code error: {:?}", e);
+            std::process::exit(1);
+        });
+
+    println!("proposal {}", proposal_id);
+}
+
+async fn propose_to_stop_canister(cmd: StopCanisterCmd, agent: Agent, proposer: NeuronId) {
+    let canister_client = GovernanceCanisterClient(NnsCanisterClient::new(
+        agent,
+        GOVERNANCE_CANISTER_ID,
+        Some(proposer),
+    ));
+
+    let canister_id = PrincipalId::try_from(cmd.canister_id).unwrap();
+    let stop_canister = StopOrStartCanister {
+        canister_id: Some(canister_id),
+        action: Some(GovernanceCanisterAction::Stop as i32),
+    };
+
+    print_proposal(&stop_canister, &cmd);
+
+    if cmd.is_dry_run() {
+        return;
+    }
+
+    let proposal_id = canister_client
+        .submit_stop_or_start_canister_proposal(
+            stop_canister,
+            cmd.url(),
+            cmd.title(),
+            cmd.summary(),
+        )
+        .await
+        .unwrap_or_else(|e| {
+            eprintln!("propose_to_stop_canister error: {:?}", e);
+            std::process::exit(1);
+        });
+
+    println!("proposal {}", proposal_id);
+}
+
+async fn propose_to_start_canister(cmd: StartCanisterCmd, agent: Agent, proposer: NeuronId) {
+    let canister_client = GovernanceCanisterClient(NnsCanisterClient::new(
+        agent,
+        GOVERNANCE_CANISTER_ID,
+        Some(proposer),
+    ));
+
+    let canister_id = PrincipalId::try_from(cmd.canister_id).unwrap();
+    let start_canister = StopOrStartCanister {
+        canister_id: Some(canister_id),
+        action: Some(GovernanceCanisterAction::Start as i32),
+    };
+
+    print_proposal(&start_canister, &cmd);
+
+    if cmd.is_dry_run() {
+        return;
+    }
+
+    let proposal_id = canister_client
+        .submit_stop_or_start_canister_proposal(
+            start_canister,
+            cmd.url(),
+            cmd.title(),
+            cmd.summary(),
+        )
+        .await
+        .unwrap_or_else(|e| {
+            eprintln!("propose_to_start_canister error: {:?}", e);
+            std::process::exit(1);
+        });
+
+    println!("proposal {}", proposal_id);
```

Comment:
NCR

Much of the code here is the same as in the other propose_to_* functions. Not sure if it is worth it, but I think we can avoid such repetition. Seems like the only "real" difference is converting from cmd to an Action type.

I think we can come up with a generic

```
fn submit_proposal_or_exit<A>(action: A, title, url, summary)
where
  Action: From<A>
-> ProposalId {
    let action = Some(Action::from(action));

    let proposal = Proposal {
        title,
        url,
        summary,
        action,
    };

    ...
}
```

by defining a bunch of

```
impl From<X> for Action
```

similar to what I did [here](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/nns/governance/src/pb/convert_struct_to_enum/manage_neuron.rs).

Reply:
Done.